src/main/java/de/tum/cit/aet/artemis/communication/service/ConversationMessagingService.java

```diff
@@ -431,4 +452,78 @@ private Conversation mayUpdateOrDeleteMessageElseThrow(Post existingMessagePost,
+     * @return returns a Page of Messages
+     */
+    public Page<Post> findMessages(PostContextFilterDTO postContextFilter, Pageable pageable, long userId) {
+        var specification = getConversationsSpecification(postContextFilter.conversationIds());
+        specification = configureSearchSpecification(specification, postContextFilter, userId);
+        // Fetch all necessary attributes to avoid lazy loading (even though relations are defined as EAGER in the domain class, specification queries do not respect this)
+        return findPostsWithSpecification(pageable, specification);
+    }
+
+    private PageImpl<Post> findPostsWithSpecification(Pageable pageable, Specification<Post> specification) {
+        // Only fetch the postIds without any left joins to avoid that Hibernate loads all objects and creates the page in Java
+        long start = System.nanoTime();
+        Page<Long> postIds = conversationMessageRepository.findPostIdsWithSpecification(specification, pageable);
+        log.debug("findPostIdsWithSpecification took {}", TimeLogUtil.formatDurationFrom(start));
+        // Fetch all necessary attributes to avoid lazy loading (even though relations are defined as EAGER in the domain class, specification queries do not respect this)
+        long start2 = System.nanoTime();
+
+        // Step 1: Fetch only the post IDs for the page
+        List<Long> ids = postIds.getContent();
+        if (ids.isEmpty()) {
+            return new PageImpl<>(List.of(), postIds.getPageable(), postIds.getTotalElements());
+        }
+        // Step 2: Fetch the base posts (with author & conversation)
+        List<Post> posts = conversationMessageRepository.findByIdIn(ids);
+        // Step 3: Fetch answer posts for those posts
+        List<AnswerPost> allAnswers = answerPostRepository.findAnswerPostsByPostIdIn(ids);
+        Map<Long, Set<AnswerPost>> answersByPostId = allAnswers.stream().collect(Collectors.groupingBy(a -> a.getPost().getId(), Collectors.toSet()));
+        // Step 4: Fetch reactions for those posts
+        List<Reaction> reactions = reactionRepository.findByPostIdIn(ids);
+        Map<Long, Set<Reaction>> reactionsByPostId = reactions.stream().collect(Collectors.groupingBy(r -> r.getPost().getId(), Collectors.toSet()));
+        // Step 5: Fetch tags for those posts
+        List<Object[]> tagsRaw = conversationMessageRepository.findTagsByPostIds(ids);
+        Map<Long, Set<String>> tagsByPostId = tagsRaw.stream().collect(Collectors.groupingBy(row -> (Long) row[0], Collectors.mapping(row -> (String) row[1], Collectors.toSet())));
+        // Step 6: Attach everything to the posts
+        for (Post post : posts) {
+            post.setAnswers(answersByPostId.getOrDefault(post.getId(), Set.of()));
+            post.setReactions(reactionsByPostId.getOrDefault(post.getId(), Set.of()));
+            post.setTags(tagsByPostId.getOrDefault(post.getId(), Set.of()));
+        }
```

Comment:
Why are we doing all of this in Java and not directly in SQL? 

Is it due to this comment `// Only fetch the postIds without any left joins to avoid that Hibernate loads all objects and creates the page in Java` ?